### PR TITLE
Add oauth client configs to DataFlowClient

### LINKED ISSUE #3617: Add security support for "scheduler-task-launcher"
As an SCDF user, I'd like to integrate the newly added `scheduler-task-launcher` module with the same security session as SCDF, so I can leverage the SSO capability to securely orchestrate the scheduled launches of Tasks.

**Acceptance:**
- Apply the same security characteristics that can be found in [tasklauncher-dataflow](https://github.com/spring-cloud-stream-app-starters/tasklauncher-dataflow) app
- Both user/pass (aka: oauth password-grant) and the typical oauth2 workflows are supported
- Verify the e2e setup with the "UAA bundled" or using any of the identity providers (eg: GH, FB etc.)
- Document the supported properties in the ref. guide.

### PR DESCRIPTION
**Pull request by jvalkeal**
- New properties client-id, client-secret, token-uri and scope in
  DataFlowClientProperties
- Enabled if `access-token` is not defined and `client-id` is set.
- Basically idea copied from a tasklauncher sink.
- With this `spring-cloud-dataflow-scheduler-task-launcher` can just get launcher with props:
  spring.cloud.scheduler.task.launcher.task-name=task1
  spring.cloud.dataflow.client.authentication.client-id=dataflow
  spring.cloud.dataflow.client.authentication.client-secret=secret
  spring.cloud.dataflow.client.authentication.token-uri=http://localhost:8080/uaa/oauth/token
  spring.cloud.dataflow.client.authentication.scope=dataflow.view,dataflow.create,dataflow.manage
- Fixes #3617

### CODE REVIEW

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientProperties.java [hunk 1]
```diff
@@ -86,6 +88,26 @@ public void setAuthentication(Authentication authentication) {
...
+
+		/**
+		 * OAuth2 Token Uri.
+		 */
+		private String tokenUri;
+
+		/**
+		 * OAuth2 Scopes.
+		 */
+		private Set<String> scope;
```
Comment (cppwfs) [inline spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientProperties.java]: nitpicky should this be scopes?

File: spring-cloud-starter-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/single/security/DataFlowClientAutoConfigurationAgaintstServerTests.java [hunk 2]
```diff
@@ -104,6 +105,30 @@ public void usingUserWithViewRoles() throws Exception {
 		context.close();
 	}
 
+	@Test
+	public void usingUserWithViewRolesWithOauth() throws Exception {
```
Comment (cppwfs) [inline spring-cloud-starter-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/single/security/DataFlowClientAutoConfigurationAgaintstServerTests.java]: `throws Exception` is not needed.

File: spring-cloud-starter-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/single/security/DataFlowClientAutoConfigurationAgaintstServerTests.java [hunk 3]
```diff
@@ -104,6 +105,30 @@ public void usingUserWithViewRoles() throws Exception {
...
+		context.register(TestApplication2.class);
+		context.refresh();
+
+		final DataFlowOperations dataFlowOperations = context.getBean(DataFlowOperations.class);
+		final AboutResource about = dataFlowOperations.aboutOperation().get();
+
+		assertNotNull(about);
+		assertEquals("myclient", about.getSecurityInfo().getUsername());
+		assertEquals(1, about.getSecurityInfo().getRoles().size());
+		context.close();
```
Comment (cppwfs) [inline spring-cloud-starter-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/single/security/DataFlowClientAutoConfigurationAgaintstServerTests.java]: Nitpicky on the test.   We should move the close context to an @After to guarantee that the context is closed.    (But that can be a different PR).

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientAutoConfiguration.java [hunk 4]
```diff
@@ -62,6 +79,14 @@ public DataFlowOperations dataFlowOperations() throws Exception{
 			template.getInterceptors().add(new OAuth2AccessTokenProvidingClientHttpRequestInterceptor(this.properties.getAuthentication().getAccessToken()));
 			logger.debug("Configured OAuth2 Access Token for accessing the Data Flow Server");
 		}
+		else if (StringUtils.hasText(this.properties.getAuthentication().getClientId())) {
+			ClientRegistration clientRegistration = clientRegistrations.findByRegistrationId("default");
```
Comment (cppwfs) [inline spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientAutoConfiguration.java]: Nitpicky `default` can be a constant.

**Comment by cppwfs** [comment]
Need for Gunnar to review it because he knows about how OAuth is configured.

**Comment by jvalkeal** [comment]
I'll wait Gunnar's comments before I polish a this pr.

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientProperties.java (hunk 1 above)
Comment (jvalkeal) [reply spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientProperties.java]: Annoyingly it's like this in spring security settings as well. Took a same key.

File: spring-cloud-starter-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/single/security/DataFlowClientAutoConfigurationAgaintstServerTests.java (hunk 2 above)
Comment (ghillert) [reply spring-cloud-starter-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/single/security/DataFlowClientAutoConfigurationAgaintstServerTests.java]: Can handle during merge

**Comment by ghillert** [comment]
LGTM - Rebased + merged.
